File: src/main/java/com/mynimef/bot/config/TelegramBot.java
package com.mynimef.bot.config;


import com.mynimef.bot.IBot;
import com.mynimef.bot.actions.IAction;
import com.mynimef.bot.actions.ICallback;
import com.mynimef.bot.actions.ISaveLogs;
import com.mynimef.bot.containers.*;
import org.telegram.telegrambots.bots.TelegramLongPollingBot;
import org.telegram.telegrambots.meta.api.methods.BotApiMethod;
import org.telegram.telegrambots.meta.api.methods.send.SendDocument;
import org.telegram.telegrambots.meta.api.methods.send.SendMessage;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage;
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageText;
import org.telegram.telegrambots.meta.api.objects.*;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.ReplyKeyboardMarkup;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton;
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.KeyboardRow;
import org.telegram.telegrambots.meta.exceptions.TelegramApiException;

import java.io.File;
import java.util.ArrayList;
import java.util.List;
import java.util.Map;


final class TelegramBot extends TelegramLongPollingBot implements IBot {

    private final Map<String, IAction> commands;
    private final IAction noCommandRecognized;
    private final Map<String, ICallback> callbacks;

    private final String username;

    private final ISaveLogs saveLog;

    TelegramBot(
            Map<String, IAction> commands,
            IAction noCommandRecognized,
            Map<String, ICallback> callbacks,
            String token,
            String username,
            ISaveLogs saveLog
    ) {
        super(token);
        this.commands = commands;
        this.noCommandRecognized = noCommandRecognized;
        this.callbacks = callbacks;
        this.username = username;
        this.saveLog = saveLog;
    }

    @Override
    public String getBotUsername() { return username; }

    @Override
    public void onUpdateReceived(Update update) {
        new Thread(() -> {
            if (update.hasMessage()) {
                Message message = update.getMessage();

                if (message.hasText()) {
                    onMessageReceived(message);
                }
            } else if (update.hasCallbackQuery()) {
                if (callbacks != null) {
                    onCallbackReceived(update.getCallbackQuery());
                } else {
                    sendMessage(
                            update.getCallbackQuery().getMessage().getChatId().toString(),
                            "There are no callbacks. Please, set them properly"
                    );
                }
            }
        }).start();
    }

    private void onMessageReceived(Message message) {
        String text = message.getText();
        Chat chat = message.getChat();
        String chatId = message.getChatId().toString();
        Integer messageId = message.getMessageId();
        String username = chat.getUserName();
        String firstName = chat.getFirstName();
        String lastName = chat.getLastName();

        if (saveLog != null) {
            saveLog.log(text, chatId, messageId, username, firstName, lastName);
        }
        IAction command = commands.get(text);
        if (command != null) {
            command.action(text, chatId, messageId, username, firstName, lastName, this);
        } else {
            noCommandRecognized.action(text, chatId, messageId, username, firstName, lastName, this);
        }
    }

    private Integer sendMessage(BotApiMethod<?> reply) {
        if (reply != null) {
            try {
                Message sentMessage = (Message) execute(reply);
                return sentMessage.getMessageId();
            } catch (TelegramApiException e) {
                e.printStackTrace();
            }
        }
        return null;
    }

    private void sendDoc(SendDocument doc) {
        try {
            execute(doc);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }
    }

    @Override
    public Integer sendMessage(String chatId, BotMessage message) {
        SendMessage sendMessage = new SendMessage();
        sendMessage.setChatId(chatId);
        sendMessage.setText(message.getText());

        if (message.doesHaveButtons()) {
            sendMessage.setReplyMarkup(setReply(message.getButtons()));
        }

        Integer messageId = sendMessage(sendMessage);
        for (BotFile file: message.getFiles()) {
            sendDoc(chatId, file);
        }
        return messageId;
    }

    @Override
    public void sendDoc(String chatId, BotFile file) {
        SendDocument doc = new SendDocument();
        doc.setChatId(chatId);
        doc.setDocument(new InputFile(new File(file.path())));
        doc.setCaption(file.description());
        sendDoc(doc);
    }

    private static InlineKeyboardMarkup setReply(List<ButtonLine> buttons) {
        InlineKeyboardMarkup replyMarkup = new InlineKeyboardMarkup();
        List<List<InlineKeyboardButton>> keyboardButtons = new ArrayList<>();

        for (ButtonLine line: buttons) {
            List<InlineKeyboardButton> keyboardButtonsRow = new ArrayList<>();
            for (Button markup: line.getLine()) {
                InlineKeyboardButton inlineKeyboardButton = new InlineKeyboardButton();
                inlineKeyboardButton.setText(markup.getLabel());
                inlineKeyboardButton.setCallbackData(markup.getCallback());
                keyboardButtonsRow.add(inlineKeyboardButton);
            }
            keyboardButtons.add(keyboardButtonsRow);
        }

        replyMarkup.setKeyboard(keyboardButtons);
        return replyMarkup;
    }

    private static ReplyKeyboardMarkup setKeyboard(ButtonKeyboardLine[] buttons) {
        ReplyKeyboardMarkup replyMarkup = new ReplyKeyboardMarkup();
        List<KeyboardRow> keyboardButtons = new ArrayList<>();

        for (ButtonKeyboardLine line: buttons) {
            KeyboardRow keyboardRow = new KeyboardRow();
            for (String text : line.getLine()) {
                keyboardRow.add(text);
                keyboardButtons.add(keyboardRow);
            }
        }

        replyMarkup.setKeyboard(keyboardButtons);
        replyMarkup.setResizeKeyboard(true);
        replyMarkup.setOneTimeKeyboard(true);
        return replyMarkup;
    }

    @Override
    public Integer sendMessage(String chatId, String text) {
        SendMessage message = new SendMessage();
        message.setChatId(chatId);
        message.setText(text);
        return sendMessage(message);
    }

    private void onCallbackReceived(CallbackQuery query) {
        String chatId = query.getMessage().getChatId().toString();
        Integer messageId = query.getMessage().getMessageId();
        String message = query.getMessage().getText();
        Chat chat = query.getMessage().getChat();
        String username = chat.getUserName();
        String firstName = chat.getFirstName();
        String lastName = chat.getLastName();


        ICallback callback = callbacks.get(query.getData());
        if (callback != null) {
            callback.callback(chatId, messageId, message, username, firstName, lastName, this);
        } else {
            sendMessage(chatId, "There are no callback \"" + query.getData() + "\"");
        }
    }

    @Override
    public void editMessage(String chatId, Integer messageId, String text) {
        EditMessageText editMessage = new EditMessageText();
        editMessage.setChatId(chatId);
        editMessage.setMessageId(messageId);
        editMessage.setText(text);
        sendMessage(editMessage);
    }

    @Override
    public void editMessage(String chatId, Integer messageId, BotMessage message) {
        EditMessageText editMessage = new EditMessageText();

        editMessage.setChatId(chatId);
        editMessage.setMessageId(messageId);
        editMessage.setText(message.getText());

        if (message.doesHaveButtons()) {
            editMessage.setReplyMarkup(setReply(message.getButtons()));
        }

        sendMessage(editMessage);
        for (BotFile file: message.getFiles()) {
            sendDoc(chatId, file);
        }
    }

    @Override
    public void deleteMessage(String chatId, Integer messageId) {
        DeleteMessage delete = new DeleteMessage();
        delete.setChatId(chatId);
        delete.setMessageId(messageId);

        try {
            execute(delete);
        } catch (TelegramApiException e) {
            e.printStackTrace();
        }
    }
}

File: src/main/kotlin/com/mynimef/bot/config/TelegramBot.kt
package com.mynimef.bot.config

import com.mynimef.bot.IBot
import com.mynimef.bot.actions.ICallback
import com.mynimef.bot.containers.*
import org.telegram.telegrambots.client.okhttp.OkHttpTelegramClient
import org.telegram.telegrambots.longpolling.util.LongPollingSingleThreadUpdateConsumer
import org.telegram.telegrambots.meta.api.methods.botapimethods.BotApiMethod
import org.telegram.telegrambots.meta.api.methods.send.SendDocument
import org.telegram.telegrambots.meta.api.methods.send.SendMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.DeleteMessage
import org.telegram.telegrambots.meta.api.methods.updatingmessages.EditMessageText
import org.telegram.telegrambots.meta.api.objects.CallbackQuery
import org.telegram.telegrambots.meta.api.objects.InputFile
import org.telegram.telegrambots.meta.api.objects.Update
import org.telegram.telegrambots.meta.api.objects.message.Message
import org.telegram.telegrambots.meta.api.objects.replykeyboard.InlineKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.ReplyKeyboardMarkup
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardButton
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.InlineKeyboardRow
import org.telegram.telegrambots.meta.api.objects.replykeyboard.buttons.KeyboardRow
import org.telegram.telegrambots.meta.exceptions.TelegramApiException
import org.telegram.telegrambots.meta.generics.TelegramClient
import java.io.Serializable


internal typealias Action = (userCommand: UserCommand, bot: IBot) -> Unit
internal typealias SaveLog = (userCommand: UserCommand) -> Unit


internal class TelegramBot(
    private val commands: Map<String, Action>,
    private val noCommandRecognized: Action,
    private val callbacks: Map<String, ICallback>?,
    token: String,
    private val saveLog: SaveLog? = null
) : LongPollingSingleThreadUpdateConsumer, IBot {

    private val telegramClient: TelegramClient = OkHttpTelegramClient(token)

    override fun consume(update: Update) = Thread {
        if (update.hasMessage()) {
            val message = update.message

            if (message.hasText()) {
                onMessageReceived(message)
            }
        } else if (update.hasCallbackQuery()) {
            if (callbacks != null) {
                onCallbackReceived(update.callbackQuery)
            } else {
                sendMessage(
                    update.callbackQuery.message.chatId.toString(),
                    "There are no callbacks. Please, set them properly"
                )
            }
        }
    }.start()

    private fun onMessageReceived(message: Message) {
        val userCommand = UserCommand(
            text = message.text,
            chatId = message.chatId.toString(),
            messageId = message.messageId,
            username = message.chat.userName,
            firstName = message.chat.firstName,
            lastName = message.chat.lastName
        )

        saveLog?.let { it(userCommand) }
        val action = commands[message.text]
        if (action != null) {
            action(userCommand, this)
        } else {
            noCommandRecognized(userCommand, this)
        }
    }

    private fun<T: Serializable> sendMessage(reply: BotApiMethod<T>): Int? {
        try {
            val sentMessage = telegramClient.execute(reply) as Message
            return sentMessage.messageId
        } catch (e: TelegramApiException) {
            e.printStackTrace()
        }
        return null
    }

    private fun sendDoc(doc: SendDocument) {
        try {
            telegramClient.execute(doc)
        } catch (e: TelegramApiException) {
            e.printStackTrace()
        }
    }

    override fun sendMessage(chatId: String, message: BotMessage): Int? {
        val sendMessage = SendMessage(
            chatId,
            message.text
        )

        if (message.buttonLines.isNotEmpty()) {
            sendMessage.replyMarkup = setButtons(message.buttonLines)
        }

        val messageId = sendMessage(sendMessage)
        for (file in message.files) {
            sendMessage(chatId, file)
        }
        return messageId
    }

    override fun sendMessage(chatId: String, botFile: BotFile) {
        val doc = SendDocument(chatId, InputFile(botFile.file))
        doc.caption = botFile.description
        sendDoc(doc)
    }

    override fun sendMessage(chatId: String, text: String): Int? {
        val message = SendMessage(
            chatId,
            text
        )
        return sendMessage(message)
    }

    private fun onCallbackReceived(query: CallbackQuery) {
        val chatId = query.message.chatId.toString()
        val messageId = query.message.messageId

        val callback = callbacks!![query.data]
        //callback?.callback(chatId, messageId, message, username, firstName, lastName, this)
        //    ?: sendMessage(chatId, "There are no callback \"" + query.data + "\"")
    }

    override fun editMessage(chatId: String, messageId: Int, text: String) {
        val editMessage = EditMessageText(text)
        editMessage.chatId = chatId
        editMessage.messageId = messageId
        sendMessage(editMessage)
    }

    override fun editMessage(chatId: String, messageId: Int, message: BotMessage) {
        val editMessage = EditMessageText(message.text)

        editMessage.chatId = chatId
        editMessage.messageId = messageId

        if (message.buttonLines.isNotEmpty()) {
            editMessage.replyMarkup = setButtons(message.buttonLines)
        }

        sendMessage(editMessage)
        for (file in message.files) {
            sendMessage(chatId, file)
        }
    }

    override fun deleteMessage(chatId: String, messageId: Int) {
        val delete = DeleteMessage(chatId, messageId)
        try {
            telegramClient.execute(delete)
        } catch (e: TelegramApiException) {
            e.printStackTrace()
        }
    }

}


private fun setButtons(buttonsLines: List<List<Button>>): InlineKeyboardMarkup {
    val keyboardButtons: MutableList<InlineKeyboardRow> = ArrayList()
    buttonsLines.forEach { buttonLine ->
        val keyboardButtonsRow = InlineKeyboardRow()
        buttonLine.forEach { button ->
            val inlineKeyboardButton = InlineKeyboardButton(button.label)
            inlineKeyboardButton.callbackData = button.callback
            keyboardButtonsRow.add(inlineKeyboardButton)
        }
        keyboardButtons.add(keyboardButtonsRow)
    }
    return InlineKeyboardMarkup(keyboardButtons)
}


private fun setKeyboard(buttons: Array<ButtonKeyboardLine>): ReplyKeyboardMarkup {
    val keyboardButtons: MutableList<KeyboardRow> = ArrayList()

    for (line in buttons) {
        val keyboardRow = KeyboardRow()
        for (text in line.line) {
            keyboardRow.add(text)
            keyboardButtons.add(keyboardRow)
        }
    }

    return ReplyKeyboardMarkup(keyboardButtons).apply {
        resizeKeyboard = true
        oneTimeKeyboard = true
    }
}
